src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -46,11 +47,16 @@ private static IEnumerable<TSource> WhereIterator<TSource>(IEnumerable<TSource>
             if (selector == null) throw Error.ArgumentNull("selector");
             Iterator<TSource> iterator = source as Iterator<TSource>;
             if (iterator != null) return iterator.Select(selector);
-            TSource[] array = source as TSource[];
-            if (array != null) return new WhereSelectArrayIterator<TSource, TResult>(array, null, selector);
-            List<TSource> list = source as List<TSource>;
-            if (list != null) return new WhereSelectListIterator<TSource, TResult>(list, null, selector);
-            return new WhereSelectEnumerableIterator<TSource, TResult>(source, null, selector);
+            IList<TSource> ilist = source as IList<TSource>;
+            if (ilist != null)
+            {
+                TSource[] array = source as TSource[];
+                if (array != null) return new SelectArrayIterator<TSource, TResult>(array, selector);
+                List<TSource> list = source as List<TSource>;
+                if (list != null) return new SelectListIterator<TSource, TResult>(list, selector);
```

Comment:
It's the IList case that's truly new here, with the others being derived from the previous code. I would probably have just done the IList case if coming at a tabla rasa, or maybe just the IList and the array, but I'm loath to remove a previous optimisation unless I'm convinced it's actually making things slower (or if it hits a buggy case, obviously).
The IList case can benefit in terms of ToArray (and there's scope for further work on the cases where an iterator can know its size in advance) with select, so I added that, but it can't with Where, so I left that as is.